### PR TITLE
server/api: add label api

### PR DESCRIPTION
**Pull request by nolouch**
for https://github.com/pingcap/pd/issues/505
PTAL @siddontang @andelf @huachaohuang

### CODE REVIEW

File: server/api/home.go [hunk 1]
```diff
@@ -32,3 +32,7 @@ func newHomeHandler(rd *render.Render) *homeHandler {
 func (h *homeHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 	h.rd.HTML(w, http.StatusOK, "index", r.Host)
 }
+
+func (h *homeHandler) Hello(w http.ResponseWriter, r *http.Request) {
```
Comment (siddontang) [inline server/api/home.go]: why need hello here?

File: server/api/label.go [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+}
+
+func (h *labelsHandler) Get(w http.ResponseWriter, r *http.Request) {
+	cluster := h.svr.GetRaftCluster()
+	if cluster == nil {
+		h.rd.JSON(w, http.StatusInternalServerError, errNotBootstrapped.Error())
+		return
+	}
+	var labels []*metapb.StoreLabel
+	m := make(map[string]bool)
```
Comment (siddontang) [inline server/api/label.go]: use `map[string]struct{}`

File: server/api/router.go [hunk 3]
```diff
@@ -51,6 +51,11 @@ func createRouter(prefix string, svr *server.Server) *mux.Router {
 	router.HandleFunc("/api/v1/store/{id}", storeHandler.Delete).Methods("DELETE")
 	router.Handle("/api/v1/stores", newStoresHandler(svr, rd)).Methods("GET")
 
+	labelsHandler := newLabelsHandler(svr, rd)
+	router.HandleFunc("/api/v1/labels", labelsHandler.Get).Methods("GET")
```
Comment (siddontang) [inline server/api/router.go]: do we have tests to cover the HTTP API?

File: server/api/label.go [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+		patternRe *regexp.Regexp
+		err       error
+	)
+
+	if pattern != "" && patternReString != "" {
+		return nil, errLabelConflictFilterRequst
+	}
+	if pattern == "" {
+		isRegexp = true
+		patternRe, err = regexp.Compile(patternReString)
```
Comment (siddontang) [inline server/api/label.go]: the empty patternReString can match everything?

File: server/api/label.go [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+		isRegexp:  isRegexp,
+		patternRe: patternRe,
+	}, nil
+}
+
+func (filter *storeLabelFilter) filterLabelKey(l *metapb.StoreLabel) bool {
+	if filter.isRegexp {
+		if filter.patternRe.MatchString(l.Key) {
+			return true
+		}
```
Comment (siddontang) [inline server/api/label.go]: should return `filter.patternRe.MatchString(l.Key)` directly here?

File: server/api/home.go (hunk 1 above)
Comment (nolouch) [inline server/api/home.go]: It used to check if pd address is working in `pd-ctl`.

File: server/api/label.go (hunk 4 above)
Comment (nolouch) [inline server/api/label.go]: yes, it will match everything

File: server/api/home.go (hunk 1 above)
Comment (siddontang) [inline server/api/home.go]: support `/ping` and return HTTP 200 is enough, no need any body.

File: server/api/router.go (hunk 3 above)
Comment (nolouch) [inline server/api/router.go]: now no test for the API of store and label, I will try to do it in one pr through mock some stores.

File: server/api/label.go [hunk 6]
```diff
@@ -0,0 +1,210 @@
...
+	"regexp"
+
+	"github.com/juju/errors"
+	"github.com/pingcap/kvproto/pkg/metapb"
+	"github.com/pingcap/pd/server"
+	"github.com/unrolled/render"
+)
+
+var (
+	errLabelConflictFilterRequst = errors.New("Cannot use string and regexp in same time")
```
Comment (huachaohuang) [inline server/api/label.go]: s/Requst/Request

File: server/api/label.go [hunk 7]
```diff
@@ -0,0 +1,210 @@
...
+	errLabelReex                 = errors.New("Cannot compile regexp, Please check it")
```
Comment (huachaohuang) [inline server/api/label.go]: s/Reex/Regex

File: server/api/label.go [hunk 8]
```diff
@@ -0,0 +1,210 @@
...
+				ret = append(ret, s)
+				break
+			}
+		}
+
+	}
+	return ret
+}
+
+type storeLabelFilter struct {
```
Comment (huachaohuang) [inline server/api/label.go]: I think the regex pattern is enough, the pattern is duplicated.

File: server/api/router.go [hunk 9]
```diff
@@ -51,6 +51,11 @@ func createRouter(prefix string, svr *server.Server) *mux.Router {
...
+	router.HandleFunc("/api/v1/labels/list", labelsHandler.List).Methods("GET")
```
Comment (huachaohuang) [inline server/api/router.go]: It's not RESTful, isn't it the same with `/api/v1/stores`?

File: server/api/label.go (hunk 8 above)
Comment (nolouch) [inline server/api/label.go]: What do you mean?just use regex?

Comment (huachaohuang) [inline server/api/label.go]: Yes.

File: server/api/router.go (hunk 9 above)
Comment (siddontang) [inline server/api/router.go]: not the same, like `/labels/stores` may be better.

**Comment by siddontang** [comment]
PTAL @overvenus

File: server/api/home.go [hunk 10]
```diff
@@ -32,3 +32,7 @@ func newHomeHandler(rd *render.Render) *homeHandler {
 func (h *homeHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 	h.rd.HTML(w, http.StatusOK, "index", r.Host)
 }
+
+func (h *homeHandler) Ping(w http.ResponseWriter, r *http.Request) {
```
Comment (overvenus) [inline server/api/home.go]: What is this for?

I considered you might be interested in [Keep-Alive](https://www.w3.org/Protocols/HTTP/1.1/draft-ietf-http-v11-spec-01#Keep-Alive).

File: server/api/label.go [hunk 11]
```diff
@@ -0,0 +1,130 @@
...
+}
+
+func (h *labelsHandler) GetStore(w http.ResponseWriter, r *http.Request) {
+	cluster := h.svr.GetRaftCluster()
+	if cluster == nil {
+		h.rd.JSON(w, http.StatusInternalServerError, errNotBootstrapped.Error())
+		return
+	}
+
+	name := r.URL.Query().Get("name")
```
Comment (overvenus) [inline server/api/label.go]: What if there is no `name`?

Comment (nolouch) [inline server/api/label.go]: it will be ""

Comment (overvenus) [inline server/api/label.go]: So, `newStoresLabelFilter` will return an error?

Comment (nolouch) [inline server/api/label.go]: no, it will match all stores

File: server/api/home.go (hunk 10 above)
Comment (nolouch) [inline server/api/home.go]: It used to check if pd address is working in pd-ctl.

**Comment by siddontang** [comment]
CI failed @nolouch

**Comment by siddontang** [comment]
Please fix the CI then merge the PR

File: server/api/home.go (hunk 10 above)
Comment (overvenus) [inline server/api/home.go]: I think member API would be better.

Comment (siddontang) [inline server/api/home.go]: both are ok, we will use ping to check the RTT #524 later.

File: server/api/label.go (hunk 11 above)
Comment (overvenus) [inline server/api/label.go]: Could you name it to `GetStores`?
```diff
-func (h *labelsHandler) GetStore(w http.ResponseWriter, r *http.Request)
+func (h *labelsHandler) GetStores(w http.ResponseWriter, r *http.Request)
```

**Pull request merged**